### PR TITLE
Merkle Tree Impl 

### LINKED ISSUE #24: Implement `MerkleTree` for the `hash` module

### PR DESCRIPTION
**Pull request by FranFiuba**
Resolves #24

### CODE REVIEW

File: crypto/src/merkle_tree/mod.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+pub mod merkle_tree;
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:1]: Newline missing also here

**Comment by ilitteri** [comment]
I'll convert this PR to draft as it is in progress.

File: crypto/src/hash/traits.rs [hunk 2]
```diff
@@ -9,10 +9,11 @@ where
...
     /// Hashes a field element into one. Also known as one to one hashing.
-    fn hash_one(input: field::element::FieldElement<F>) -> field::element::FieldElement<F>;
+    fn hash_one(&self, input: field::element::FieldElement<F>) -> field::element::FieldElement<F>;
 
     /// Hashes two field elements into one. Also known as two to one hashing.
     fn hash_two(
+        &self,
         left: field::element::FieldElement<F>,
         right: field::element::FieldElement<F>,
     ) -> field::element::FieldElement<F>;
```
Comment (ilitteri) [inline crypto/src/hash/traits.rs:19]: The changes in this file belong to another PR. Merge `main` into this branch to fix the diff.

File: crypto/src/merkle_tree/merkle_tree.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+impl<F: IsField, H: IsCryptoHash<F>> MerkleTree<F, H> {
+    pub fn build(values: &[FieldElement<F>], hasher: H) -> MerkleTree<F, H> {
+        let mut leafs: Vec<TreeNode<F>> = hash_leafs(values, &hasher);
+        let mut level: Vec<TreeNode<F>> = hash_level(&mut leafs, &hasher);
+
+        while level.len() > 1 {
+            level = hash_level(&mut level, &hasher);
+        }
+
+        return MerkleTree {
```
Comment (ilitteri) [inline crypto/src/merkle_tree/merkle_tree.rs]: This `return` is not needed.

File: crypto/src/hash/traits.rs (hunk 2 above)
Comment (FranFiuba) [reply crypto/src/hash/traits.rs]: fixed

File: crypto/src/merkle_tree/merkle_tree.rs (hunk 3 above)
Comment (FranFiuba) [reply crypto/src/merkle_tree/merkle_tree.rs]: yes fixed those kind of warnings in the following commit

**Comment by ilitteri** [comment]
The PR is looking good, the `clippy` job is not passing, you can use `make clippy` to reproduce the CI's output.

File: crypto/src/merkle_tree/mod.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+pub struct Proof<F: IsField, H: IsCryptoHash<F>> {
+    value: FieldElement<F>,
+    merkle_path: Vec<TreeNode<F>>,
+    hasher: H,
+}
+
+impl<F: IsField, H: IsCryptoHash<F>> Proof<F, H> {
+    pub fn verify(self, root_hash: FieldElement<F>) -> bool {
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:114]: I think that it would be better to add a method in the `MerkleTree` implementation like `verify(proof: Proof<F, H>, root: FieldElement<F>) -> bool;` so then you can call it like `MerkleTree::verify`. This way the function is not related directly with the proof itself.

File: crypto/src/merkle_tree/mod.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                hashed_value = self
+                    .hasher
+                    .hash_two(hashed_value, sibiling.borrow().hash.clone());
+            }
+        }
+        hashed_value == root_hash
+    }
+}
+
+type TreeNode<F> = Rc<RefCell<Node<F>>>;
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:128]: Does this work if we remove `Rc` and `RefCell`?

File: crypto/src/merkle_tree/mod.rs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    pub fn get_proof(&self, value: FieldElement<F>) -> Option<Proof<F, H>> {
+        let hashed_value = self.hasher.hash_one(value.clone());
+
+        if let Some(leaf) = self
+            .leafs
+            .iter()
+            .find(|node| node.borrow().hash == hashed_value)
+        {
+            let mut merkle_path: Vec<TreeNode<F>> = Vec::new();
+            merkle_path = build_merkle_path(leaf.clone(), &mut merkle_path).to_vec();
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:40]: These lines could be collapsed like this (if I'm not mistaken)
```suggestion
            let merkle_path = build_merkle_path(leaf.clone(), &mut Vec::new()).to_vec();
```

File: crypto/src/merkle_tree/mod.rs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    right.borrow_mut().sibiling = Some(left);
+    right.borrow_mut().parent = Some(parent.clone());
+
+    parent
+}
+
+fn build_merkle_path<F: IsField>(
+    node: TreeNode<F>,
+    merkle_path: &mut Vec<TreeNode<F>>,
+) -> &Vec<TreeNode<F>> {
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:110]: Could we return a `Vec<TreeNode<F>>` instead of its reference? With this change you remove the need for the caller of this function to use the `to_vec()` method.

File: crypto/src/merkle_tree/mod.rs [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+pub struct Proof<F: IsField, H: IsCryptoHash<F>> {
+    value: FieldElement<F>,
+    merkle_path: Vec<TreeNode<F>>,
+    hasher: H,
+}
+
+type TreeNode<F> = Rc<RefCell<Node<F>>>;
+
+fn build_tree_node<F: IsField>(hash: FieldElement<F>) -> TreeNode<F> {
```
Comment (ilitteri) [inline crypto/src/merkle_tree/mod.rs:127]: This function could be the method `Node::new`.

**Review by ilitteri — APPROVED** [review]
LGTM, nice work! We can merge after documenting.

File: crypto/src/merkle_tree/mod.rs [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    merkle_path.to_vec()
+}
+
+pub struct Proof<F: IsField, H: IsCryptoHash<F>> {
+    value: FieldElement<F>,
+    merkle_path: Vec<TreeNode<F>>,
+    hasher: H,
+}
+
+type TreeNode<F> = Rc<RefCell<Node<F>>>;
```
Comment (pablodeymo) [inline crypto/src/merkle_tree/mod.rs:125]: why do you need a RefCell?

Comment (FranFiuba) [reply crypto/src/merkle_tree/mod.rs:125]: We agreed to merge this version, so that it is possible to use the merkle tree and after that I would open an other pr and refactor this implementation into a simpler version.

**Review by pablodeymo — APPROVED** [review]
As we talked with @FranFiuba , we can merge this PR and then refactor the code with a simpler version.
